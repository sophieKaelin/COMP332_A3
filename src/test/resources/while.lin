//
// This file is part of COMP332 Assignment 3 2019.
//
// Lintilla, a simple functional programming language.
//
// Â© 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// A while loop in functional style.
//

fn while(p : fn(int)->bool, f : fn(int)->int) -> (fn(int)->int) {

    fn while_aux(n : int) -> int {
        if p(n) {
            while_aux(f(n))
        } else {
            n
        }
    };

    while_aux
};

fn step(n : int) -> int {
    print(n);
    n * 3 + 1
};

fn less_10000(n : int) -> bool {
    n < 10000
};

let funny = while(less_10000, step);

print funny(2)
