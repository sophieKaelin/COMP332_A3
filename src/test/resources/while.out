Program tree:
-------------

Program(
    Vector(
        FnDecl(
            IdnDef("while"),
            Vector(
                ParamDecl(
                    IdnDef("p"),
                    FnType(Vector(IntType()), BoolType())),
                ParamDecl(IdnDef("f"), FnType(Vector(IntType()), IntType()))),
            Some(FnType(Vector(IntType()), IntType())),
            Block(
                Vector(
                    FnDecl(
                        IdnDef("while_aux"),
                        Vector(ParamDecl(IdnDef("n"), IntType())),
                        Some(IntType()),
                        Block(
                            Vector(
                                IfExp(
                                    AppExp(
                                        IdnExp(IdnUse("p")),
                                        Vector(IdnExp(IdnUse("n")))),
                                    Block(
                                        Vector(
                                            AppExp(
                                                IdnExp(IdnUse("while_aux")),
                                                Vector(
                                                    AppExp(
                                                        IdnExp(IdnUse("f")),
                                                        Vector(
                                                            IdnExp(
                                                                IdnUse("n")))))))),
                                    Block(Vector(IdnExp(IdnUse("n")))))))),
                    IdnExp(IdnUse("while_aux"))))),
        FnDecl(
            IdnDef("step"),
            Vector(ParamDecl(IdnDef("n"), IntType())),
            Some(IntType()),
            Block(
                Vector(
                    PrintExp(IdnExp(IdnUse("n"))),
                    PlusExp(
                        StarExp(IdnExp(IdnUse("n")), IntExp(3)),
                        IntExp(1))))),
        FnDecl(
            IdnDef("less_10000"),
            Vector(ParamDecl(IdnDef("n"), IntType())),
            Some(BoolType()),
            Block(Vector(LessExp(IdnExp(IdnUse("n")), IntExp(10000))))),
        LetDecl(
            IdnDef("funny"),
            AppExp(
                IdnExp(IdnUse("while")),
                Vector(IdnExp(IdnUse("less_10000")), IdnExp(IdnUse("step"))))),
        PrintExp(AppExp(IdnExp(IdnUse("funny")), Vector(IntExp(2))))))

Environment out:
----------------

scope
    "funny" -> Variable(AppExp(IdnExp(IdnUse(while)),Vector(IdnExp(IdnUse(less_10000)), IdnExp(IdnUse(step)))))
    "step" -> Function(fn(int)->int,Block(Vector(PrintExp(IdnExp(IdnUse(n))), PlusExp(StarExp(IdnExp(IdnUse(n)),IntExp(3)),IntExp(1)))))
    "less_10000" -> Function(fn(int)->bool,Block(Vector(LessExp(IdnExp(IdnUse(n)),IntExp(10000)))))
    "while" -> Function(fn(fn(int)->bool,fn(int)->int)->(fn(int)->int),Block(Vector(FnDecl(IdnDef(while_aux),Vector(ParamDecl(IdnDef(n),int)),Some(int),Block(Vector(IfExp(AppExp(IdnExp(IdnUse(p)),Vector(IdnExp(IdnUse(n)))),Block(Vector(AppExp(IdnExp(IdnUse(while_aux)),Vector(AppExp(IdnExp(IdnUse(f)),Vector(IdnExp(IdnUse(n)))))))),Block(Vector(IdnExp(IdnUse(n)))))))), IdnExp(IdnUse(while_aux)))))

Target tree:
------------

List(
    IClosure(
        Some("while"),
        List("p", "f"),
        List(
            IClosure(
                Some("while_aux"),
                List("n"),
                List(
                    IVar("n"),
                    IVar("p"),
                    ICall(),
                    IBranch(
                        List(
                            IVar("n"),
                            IVar("f"),
                            ICall(),
                            IVar("while_aux"),
                            ICall()),
                        List(IVar("n"))))),
            IClosure(None, List("while_aux"), List(IVar("while_aux"))),
            ICall())),
    IClosure(
        None,
        List("while"),
        List(
            IClosure(
                Some("step"),
                List("n"),
                List(
                    IVar("n"),
                    IPrint(),
                    IVar("n"),
                    IInt(3),
                    IMul(),
                    IInt(1),
                    IAdd())),
            IClosure(
                None,
                List("step"),
                List(
                    IClosure(
                        Some("less_10000"),
                        List("n"),
                        List(IVar("n"), IInt(10000), ILess())),
                    IClosure(
                        None,
                        List("less_10000"),
                        List(
                            IVar("less_10000"),
                            IVar("step"),
                            IVar("while"),
                            ICall(),
                            IClosure(
                                None,
                                List("funny"),
                                List(
                                    IInt(2),
                                    IVar("funny"),
                                    ICall(),
                                    IPrint())),
                            ICall())),
                    ICall())),
            ICall())),
    ICall())

Output:
-------

2
7
22
67
202
607
1822
5467
16402
